kv: add TenantInfo rpc to resolve tenant name to ID (do_not_commit)

cthumuluru-crdb opened the pull request:
Add a new TenantInfo rpc to internal server that takes a tenant name and returns
tenant info for that tenant. Besides other information, the response includes
TenantID. TenantInfo rpc is accessible to secondary tenants. When tenant name
is invalid, we throw internal error instead of tenant not found.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/140463)
<!-- Reviewable:end -->

stevendanna commented on pkg/server/node.go:
```diff
@@ -2979,3 +2980,30 @@ func (n *Node) GetRangeDescriptors(
+	ctx context.Context, req *kvpb.TenantInfoRequest,
+) (*kvpb.TenantInfoResponse, error) {
+	ex := n.execCfg.InternalDB.Executor()
+	query := fmt.Sprintf("SELECT id FROM system.tenants WHERE name = '%s'", req.TenantName)
```
Let's definitely use a query parameter here rather than Sprintf.

stevendanna commented on pkg/server/node.go:
```diff
@@ -2979,3 +2980,30 @@ func (n *Node) GetRangeDescriptors(
 	})
 }
+
+func (n *Node) TenantInfo(
```
As discussed, I _kinda_ expected this to live on the admin server.

We should think about authorization here.  A tenant should only be able to request the info for the tenant it is authorized as.

Ahh, I see now that you have a TODO for this elsehwere.

shaikzakiriitm reviewed (commented):
> Reviewed 1 of 5 files at r6, 10 of 10 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140463)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @cthumuluru-crdb)

___
*[`pkg/kv/kvclient/kvtenant/connector.go` line 555 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/140463#-OItQVZV800Rfyzxahmm:-OItQVZV800Rfyzxahmn:b-4bs490) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f973081932c9c99e551ca4a60b8567f05da66fb/pkg/kv/kvclient/kvtenant/connector.go#L555)):*
> ```Go
> 			log.Warningf(ctx, "error resolving tenant name [%s] to ID", tenantName)
> 			if grpcutil.IsAuthError(err) {
> 				return roachpb.TenantID{}, err
> ```

Nit: why can't we just return nil instead of empty tenantid? 
___
*[`pkg/server/admin.go` line 3952 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/140463#-OItc3Yf4IeCqO0OkvMF:-OItc3Yf4IeCqO0OkvMG:b3h1fox) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f973081932c9c99e551ca4a60b8567f05da66fb/pkg/server/admin.go#L3952)):*
> ```Go
> 	}
> 
> 	getTenantInfoQuery := `SELECT id, name
> ```

Nit: We already have access to the `name` field. 



_Suggestion:_
```Go
	getTenantInfoQuery := `SELECT id
```
___
*[`pkg/server/admin.go` line 3967 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/140463#-OItauX79CoCOgp44Aq8:-OItauX79CoCOgp44Aq9:b-custi7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f973081932c9c99e551ca4a60b8567f05da66fb/pkg/server/admin.go#L3967)):*
> ```Go
> 				TenantId: &roachpb.TenantID{},
> 			},
> 		}, nil
> ```

shouldn't we return an error like the `HTTP 404` here, indicating that the requested tenant name wasn't present in the list of tenants? 


<!-- Sent from Reviewable.io -->

cthumuluru-crdb committed:
cli, server: add cli flags to `mt start-sql` to pass tenant name.

This PR adds a couple of new CLI flags `--tenant-name` and `tenant-name-file`
to pass tenant name. Currently `./cockroach mt start-sql` command supports
`tenant-id` and `tenant-id-file` to pass tenant ID under which the SQL server
shuold be started. The tenant name flags intend to replace tenant ID flags
from the CLI commands.

`tenant-id-file` enables the initialization of a warm pool of SQL pods, with
the option to delay their binding to a tenant.

Epic: CRDB-39093
Informs: https://github.com/cockroachdb/cockroach/issues/105340, https://github.com/cockroachdb/cockroach/issues/105339
Release note: None

cthumuluru-crdb committed:
cli: unhide `--tenant-name-scope` flag

`--tenant-name-scope` cli flag is used to generate tenant scoped
client certificates using tenant names. `--tenant-id-scope` cli
flag serves the same purpose but using tenant IDs.

`--tenant-name-scope` offers better user experience compared to
`--tenant-id-scope` and will eventually replace it.

Epic: CRDB-39093
Informs: #105340, #105339
Release note: None

cthumuluru-crdb committed:
roachpb: add TenantIdentifier to wrap TenantID and TenantName

Introduce TenantIdentifier interface to wrap TenantID and TenantName. This
change makes it easy to transition away from TenantID in user-visible surfaces.

Epic: CRDB-39093
Informs: #105340, #105339
Release note: None
